### PR TITLE
residue_partition.py refactor pt.1

### PR DESCRIPTION
**Pull request by matthew7838**
### Summary
This PR is the first part of refactoring `residue_partition.py` of the Soil module from "pseudocode_soil" S.6.B.I. As for this part, the PR cover methods from S.6.B.I.1 to S.6.B.I.5 and S.6.B.I.7. All methods created are static methods and will be used later on in the main routine to update attributes that will be added in` Layer `and `Soil` class.

Every method is tested covering possibly most edge cases.

### Changes

- All methods in this part are static
- Attribute and method names were changed to be more descriptive
- As for the current PR, there will be no attributes being modified directly

###  Other changes

-  Fixed some errors in the original code that contradicts the pseudocode_soil according to @morrowcj 's clarification, expected more errors in later parts to need to be changed based on other clarifications

**Note**
In S.6.B.I.5, AGmet active decomp contradicts the code again. It's currently following the number given in the psudocode_soil file, it will require future clarification and has been added as TODO, not sure if issues should be created.

### CODE REVIEW

File: SC_redesign/Crop_and_Soil/soil/carbon_cycling/residue_partition.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
+from typing import Optional
+from SC_redesign.Crop_and_Soil.soil.soil_data import SoilData
+
+"""
+This class contains all necessary methods that involves residue partition, including both plant and soil and also
```
Comment (Isgotkowitz) [inline SC_redesign/Crop_and_Soil/soil/carbon_cycling/residue_partition.py]:
```suggestion
This class contains all necessary methods that involve residue partition, including both plant and soil and also
```

File: SC_redesign/Crop_and_Soil/soil/carbon_cycling/residue_partition.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+        # TODO nitrogen_fraction_plant_residue calculate in RuFaS [C.5.B.1] but not "accurate" for carbon use -
+        #  GitHub Issue #163
+        """This method calculates the plant lignin to nitrogen ratio when nitrogen in plant residue at harvest
+        is greater than zero
+
+        Parameters
+        ----------
+        plant_residue_lignin_composition : float
+            lignin fraction of plant residue (unitless)
+        nitrogen_fraction_plant_residue : float
```
Comment (Isgotkowitz) [inline SC_redesign/Crop_and_Soil/soil/carbon_cycling/residue_partition.py]:
```suggestion
        nitrogen_fraction_plant_residue : float, default=0.4
```

File: SC_redesign/Crop_and_Soil/soil/carbon_cycling/residue_partition.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+            plant lignin to nitrogen ratio (Dmnl)
+
+        References
+        -------
+        pseudocode_soil S.6.B.I.2
+        """
+        if nitrogen_fraction_plant_residue != 0:
+            return (plant_residue_lignin_composition / 100) / nitrogen_fraction_plant_residue
+        else:
+            return 0
```
Comment (Isgotkowitz) [inline SC_redesign/Crop_and_Soil/soil/carbon_cycling/residue_partition.py]: I think if you are going to do input validation here it should be more exhaustive. More specifically, checking that the nitrogen ratio is > 0 and < 1.0 (or whatever the appropriate bounds are)

File: SC_redesign/Crop_and_Soil/soil/carbon_cycling/residue_partition.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        return 0.85 - 0.18 * plant_lignin_nitrogen_ratio
+
+    @staticmethod
+    def _determine_plant_metabolic_carbon_amount(plant_metabolic_carbon_amount: float,
+                                                 plant_residue_metabolic_fraction: float,
+                                                 plant_dry_matter_residue_amount: float,
+                                                 plant_metabolic_active_carbon_usage: float,
+                                                 plant_metabolic_to_soil_carbon_amount: float) -> float:
+        """This method calculates a return the updated plant metabolic carbon amount after adding the metabolic carbon
+        in dry matter at harvest and and reduced my the amount that's decomposed and incorporated
```
Comment (Isgotkowitz) [inline SC_redesign/Crop_and_Soil/soil/carbon_cycling/residue_partition.py]:
```suggestion
        """This method calculates the updated plant metabolic carbon amount after adding the metabolic carbon
        in dry matter at harvest and and reduced by the amount that's decomposed and incorporated
```

File: SC_redesign/Crop_and_Soil/tests_SC/soil_tests/test_residue_partition.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    35.8,  # arbitrary
+])
+def test_determine_metabolic_plant_residue_ratio(plant_residue_lignin_composition: float) -> None:
+    """Test that metabolic plant residue ration is correctly determined under current nitrogen_fraction_plant_residue
+    """
+    nitrogen_fraction_plant_residue = 0.4
+    if nitrogen_fraction_plant_residue != 0:
+        expected = (plant_residue_lignin_composition / 100) / nitrogen_fraction_plant_residue
+    else:
+        expected = 0
```
Comment (Isgotkowitz) [inline SC_redesign/Crop_and_Soil/tests_SC/soil_tests/test_residue_partition.py]: If `nitrogen_fraction_plant_residue` is hardcoded, then the if-else block is unnecessary. That said I think it should be kept and different values of `nitrogen_fraction_plant_residue` should be tested, even though it has a hard-coded default in the method.

**Review by Isgotkowitz — APPROVED** [review]
Good PR Matthew. One comment is about input validation, one is about testing coverage, and the rest are typo corrections.

**Review by morrowcj — APPROVED** [review]
LGTM, after you address Ed's comments.

**Comment by morrowcj** [comment]
> **Note** In S.6.B.I.5, AGmet active decomp contradicts the code again. It's currently following the number given in the psudocode_soil file, it will require future clarification and has been added as TODO, not sure if issues should be created.

@matthew7838, yes go ahead and try to match the pseudocode for now. And yes, create issues for anything that you don't address now.

File: SC_redesign/Crop_and_Soil/soil/carbon_cycling/residue_partition.py (hunk 3 above)
Comment (matthew7838) [reply SC_redesign/Crop_and_Soil/soil/carbon_cycling/residue_partition.py]: This is following the pseudo_soil equation, not sure if I should change it. Might be something to discuss with Hector? P

File: SC_redesign/Crop_and_Soil/soil/carbon_cycling/residue_partition.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+        -------
+        pseudocode_soil S.6.B.I.2
+        """
+        if 0 < nitrogen_fraction_plant_residue < 1.0:
+            return (plant_residue_lignin_composition / 100) / nitrogen_fraction_plant_residue
+        elif nitrogen_fraction_plant_residue == 0:
+            return 0
+        else:
+            raise ValueError("Expected nitrogen_fraction_plant_residue be between 0.0-1.0, received "
+                             + str(nitrogen_fraction_plant_residue))
```
Comment (Isgotkowitz) [inline SC_redesign/Crop_and_Soil/soil/carbon_cycling/residue_partition.py:74]: This is excellent error handling

File: SC_redesign/Crop_and_Soil/soil/carbon_cycling/residue_partition.py [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+
+    @staticmethod
+    def _determine_plant_residue_metabolic_fraction(plant_lignin_nitrogen_ratio: float) -> float:
+        """This method calculates the fraction of plant residue that is metabolic
+
+        Parameters
+        ----------
+        plant_lignin_nitrogen_ratio : float
+            plant lignin to nitrogen ratio (Dmnl)
```
Comment (Isgotkowitz) [inline SC_redesign/Crop_and_Soil/soil/carbon_cycling/residue_partition.py]: What is that unit?

**Review by Isgotkowitz — APPROVED** [review]
Nice PR. The error handling (and testing of that handling) is really good.

**Pull request merged**